Waymo Open Sim Agent Challenge (WOSAC) evaluation

daphne-cornelisse opened the pull request:
## Description

- Add option to run [WOSAC](https://waymo.com/open/challenges/2025/sim-agents/) realism evaluation with:
```bash
puffer eval puffer_drive --wosac.enabled True
```

## Links
- [WOSAC 2025 configs](https://github.com/waymo-research/waymo-open-dataset/blob/master/src/waymo_open_dataset/wdl_limited/sim_agents_metrics/challenge_2025_sim_agents_config.textproto)
- [WOSAC main metrics entry point](https://github.com/waymo-research/waymo-open-dataset/blob/master/src/waymo_open_dataset/wdl_limited/sim_agents_metrics/metrics.py)


## Tasks

- [x] Basic WOSAC evaluator pipeline and integration with `puffer eval`
- [x] Collect simulated trajectories for a given policy
- [x] Add object ids to drive struct
- [x] https://github.com/Emerge-Lab/PufferDrive/pull/86
- [x] https://github.com/Emerge-Lab/PufferDrive/pull/84
- [x] Store scenario ID, and metadata (tracks to predict, sdc_index) as part of map binaries and load in env
- [x] Add initialization mode to bypass standard initialization rules, read in agents directly from metadata flags
- [x] Figure out how num_agents works and this Serial vecenv
- [x] Process and store GT WOMD data via HF? 
- [x] Implement `collect_ground_truth_data()`
- [x] Implement `compute_metrics()`

daphne-cornelisse committed:
WOSAC structure

daphne-cornelisse committed:
WOSAC structure

daphne-cornelisse committed:
WOSAC in eval

daphne-cornelisse committed:
WOSAC in eval

daphne-cornelisse committed:
Add in collect_simulated_trajectories() functionality and visualization dashboard.

daphne-cornelisse committed:
Add in collect_simulated_trajectories() functionality and visualization dashboard.

daphne-cornelisse committed:
Add in WOMD object ids.

daphne-cornelisse committed:
Add in WOMD object ids.

daphne-cornelisse committed:
Merge branch 'main' of https://github.com/Emerge-Lab/PufferDrive into dc/wosac_eval

daphne-cornelisse committed:
Load scenario id and metadata into map binary.

daphne-cornelisse committed:
Load scenario id and metadata into map binary.

daphne-cornelisse committed:
Store scenario ids in rollout dict.

daphne-cornelisse committed:
Store scenario ids in rollout dict.

daphne-cornelisse committed:
Naming

daphne-cornelisse committed:
Naming

daphne-cornelisse committed:
Merge branch 'main' of https://github.com/Emerge-Lab/PufferDrive into dc/wosac_eval

daphne-cornelisse committed:
Eval notebook

daphne-cornelisse committed:
Eval notebook

daphne-cornelisse committed:
Fix merge conflict.

daphne-cornelisse committed:
Fix merge conflict.

daphne-cornelisse committed:
Realism notebook wip

daphne-cornelisse committed:
Realism notebook wip

daphne-cornelisse committed:
Implement several WOSAC metrics.

daphne-cornelisse committed:
Implement several WOSAC metrics.

daphne-cornelisse committed:
Merge branch 'main' of https://github.com/Emerge-Lab/PufferDrive into dc/wosac_eval

daphne-cornelisse committed:
init_mode placeholder

daphne-cornelisse committed:
init_mode placeholder

daphne-cornelisse committed:
Delete messy code. Refactor with init_mode.

daphne-cornelisse committed:
Delete messy code. Refactor with init_mode.

daphne-cornelisse committed:
Simplify: rmv control_non_vehicles and just make it part of the init_mode.

daphne-cornelisse committed:
Simplify: rmv control_non_vehicles and just make it part of the init_mode.

daphne-cornelisse committed:
Rename stuff

daphne-cornelisse committed:
Rename stuff

daphne-cornelisse committed:
Merge branch 'main' of https://github.com/Emerge-Lab/PufferDrive into dc/wosac_eval

daphne-cornelisse committed:
Fix visualizer.

daphne-cornelisse committed:
Fix visualizer.

daphne-cornelisse committed:
Introduce control_tracks_to_predict as init_mode.

daphne-cornelisse committed:
Introduce control_tracks_to_predict as init_mode.

daphne-cornelisse committed:
Draw womd track indices for init_mode control_tracks_to_predict

daphne-cornelisse committed:
Draw womd track indices for init_mode control_tracks_to_predict

daphne-cornelisse committed:
Fix collect_gt_trajectories() code.

daphne-cornelisse committed:
Fix collect_gt_trajectories() code.

daphne-cornelisse committed:
Eval v1

daphne-cornelisse committed:
Eval v1

daphne-cornelisse committed:
Merge branch 'main' of https://github.com/Emerge-Lab/PufferDrive into dc/wosac_eval

daphne-cornelisse committed:
Working end-to-end WOSAC pipeline v1.

daphne-cornelisse committed:
Working end-to-end WOSAC pipeline v1.

daphne-cornelisse committed:
Stable eval without scenario_id.

daphne-cornelisse committed:
Stable eval without scenario_id.

daphne-cornelisse committed:
Add unique scenario id to group metrics.

daphne-cornelisse committed:
Add unique scenario id to group metrics.

daphne-cornelisse committed:
Cleanup.

daphne-cornelisse committed:
Cleanup.

daphne-cornelisse committed:
Add histogram estimator.

daphne-cornelisse committed:
Add histogram estimator.

daphne-cornelisse committed:
Add in likelihood.

daphne-cornelisse committed:
Add in likelihood.

daphne-cornelisse committed:
num_maps 2

daphne-cornelisse committed:
num_maps 2

daphne-cornelisse committed:
Fix histogram estimation function.

daphne-cornelisse committed:
Fix histogram estimation function.

daphne-cornelisse committed:
Fix wosac init mode bug. Enables wosac eval on any scene.

daphne-cornelisse committed:
Fix wosac init mode bug. Enables wosac eval on any scene.

daphne-cornelisse committed:
Small formatting improvement.

daphne-cornelisse committed:
Small formatting improvement.

daphne-cornelisse committed:
Rename _car_ instances to _agent_.

daphne-cornelisse committed:
Rename _car_ instances to _agent_.

daphne-cornelisse committed:
Refactor set_active_agent() function.

daphne-cornelisse committed:
Refactor set_active_agent() function.

daphne-cornelisse committed:
Minor

daphne-cornelisse committed:
Minor

daphne-cornelisse committed:
Add new map

daphne-cornelisse committed:
Add new map

daphne-cornelisse committed:
Correctly index envs to extract information.

daphne-cornelisse committed:
Correctly index envs to extract information.

Commit:
Include all kinematics metrics and realism metametric

Commit:
Include all kinematics metrics and realism metametric

daphne-cornelisse committed:
Cleanup

daphne-cornelisse committed:
Cleanup

eugenevinitsky commented on pufferlib/config/ocean/drive.ini:
```diff
@@ -50,7 +49,7 @@ adam_beta1 = 0.9
 adam_eps = 1e-8
 clip_coef = 0.2
-ent_coef = 0.001
+ent_coef = 0.005
```
this should probably not go in this PR

daphne-cornelisse committed:
Merge branch 'dc/wosac_eval' of https://github.com/Emerge-Lab/PufferDrive into dc/wosac_eval

daphne-cornelisse committed:
Fix default settings.

daphne-cornelisse committed:
Delete trajectory_comparison.png

daphne-cornelisse committed:
Fix merge conflicts.

daphne-cornelisse committed:
Refactor.

daphne-cornelisse committed:
Cleanup

daphne-cornelisse committed:
Fix missing structs.

daphne-cornelisse committed:
Fix readme

daphne-cornelisse committed:
Small improvement

daphne-cornelisse committed:
Small improvement

daphne-cornelisse committed:
Small visualizer fix.

daphne-cornelisse committed:
Delete old code.

daphne-cornelisse committed:
Small eval fixes.

daphne-cornelisse committed:
Update readme docs

Commit:
- Fixed the assert so puffer eval still works when woasc_enabled is set to
False.

- Added code for not treating the timesteps independently. You can
change the behavior in wosac.ini config file.

- Added a num_agents column to the scene-level results of WOSAC
evaluation.

Commit:
Merge branch 'dc/wosac_eval' of github.com:Emerge-Lab/PufferDrive into dc/wosac_eval

daphne-cornelisse committed:
Update readme.

daphne-cornelisse committed:
Update readme.

daphne-cornelisse committed:
Update readme.

Commit:
Added a script to sanity check the implementation of the WOSAC metrics.

Can also be used to derive empirical upper bounds on the metrics, by
computing them for the logged oracle.

Commit:
Correct initialization for metrics_sanity_check

daphne-cornelisse committed:
Update goal behavior.

daphne-cornelisse committed:
Merge branch 'dc/wosac_eval' of https://github.com/Emerge-Lab/PufferDrive into dc/wosac_eval

daphne-cornelisse committed:
Add notes

daphne-cornelisse commented on pufferlib/ocean/benchmark/estimators.py:
```diff
@@ -0,0 +1,160 @@
+
+    Args:
+        log_samples: Shape (n_agents, sample_size) - samples to evaluate
+        sim_samples: Shape (n_agents, sample_size) - samples to build distribution from
```
Correct, they will be different

daphne-cornelisse committed:
Final cleanup..

daphne-cornelisse committed:
Final cleanup..

daphne-cornelisse committed:
Leaving a comment for future me.

daphne-cornelisse commented on pufferlib/config/ocean/drive.ini:
```diff
@@ -36,18 +32,22 @@ reward_goal = 1.0
+; Options: "control_vehicles", "control_agents", "control_tracks_to_predict"
+control_mode = "control_vehicles"
+; Options: "created_all_valid", "create_only_controlled"
+init_mode = "create_all_valid"
```
Yes! Wrote docs

daphne-cornelisse committed:
Increase limit for tests.
